### PR TITLE
feat(microsoft365): Add intelligent meeting scheduling and room management

### PR DESCRIPTION
**Pull request by Shubhank-Jonnada**
## Summary

Adds intelligent meeting scheduling and room management capabilities to the Microsoft 365 integration, enabling users to find optimal meeting times, check attendee availability, discover meeting rooms, and verify room availability through Microsoft Graph API.

## Motivation and Context

Organizations need efficient tools to schedule meetings that work for all attendees while securing appropriate meeting spaces. This enhancement addresses the common challenge of coordinating schedules across multiple participants and finding available conference rooms, reducing the back-and-forth typically required for meeting planning.

## Technical Approach

Implements four new actions using Microsoft Graph API endpoints:
- **findMeetingTimes API** for intelligent scheduling with confidence scoring
- **getSchedule API** for free/busy availability lookups  
- **Places API** for comprehensive room discovery and management
- Enhanced OAuth2 permissions for Schedule.Read.All and Place.Read.All scopes

The implementation includes robust error handling, Microsoft Graph timestamp parsing (handling 7-digit fractional seconds), and automatic working hours intelligence that respects user timezone and Outlook configurations.

## Type of Change

- [x] Feature (non-breaking change that adds functionality)
- [ ] Bug fix (non-breaking change that resolves an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactor (code change that neither fixes a bug nor adds a feature)
- [ ] Documentation update
- [ ] Performance improvement
- [ ] Test coverage

## Changes

- Added `find_meeting_times` action for intelligent meeting scheduling with attendee availability analysis and confidence scoring
- Added `get_schedule` action for free/busy availability lookups with working hours and detailed schedule items
- Added `list_rooms` and `check_room_availability` actions for comprehensive meeting room discovery and conflict detection
- Updated OAuth2 permissions to include Schedule.Read.All and Place.Read.All scopes for scheduling and room management
- Enhanced Microsoft Graph timestamp parsing to handle 7-digit fractional seconds compatibility
- Added comprehensive test coverage with 8 new test methods covering success, error, and edge case scenarios
- Updated SDK version requirement to autohive-integrations-sdk~=1.0.2
- Added extensive documentation with usage examples for all new scheduling workflows
- Cleaned up __init__.py formatting and removed unused variables

## Checklist

- [x] Code compiles without errors
- [x] Changes have been self-reviewed
- [x] Unit tests added or updated where applicable
- [x] Code follows project style guidelines
- [x] Documentation updated as needed

### CODE REVIEW

**Comment by Sagsgit** [comment]
@codex review

**Comment by Sagsgit** [comment]
@codex

File: microsoft365/microsoft365.py [hunk 1]
```diff
@@ -2210,4 +2210,430 @@ async def execute(self, inputs: Dict[str, Any], context: ExecutionContext) -> Ac
...
+
+            start_dt = inputs.get("start_datetime")
+            end_dt = inputs.get("end_datetime")
+
+            # Always apply defaults as documented in schema
+            if not start_dt:
+                start_parsed = datetime.utcnow()
+                start_dt = start_parsed.isoformat() + "Z"
+            else:
+                start_parsed = datetime.fromisoformat(start_dt.replace("Z", ""))
```
[the remark this answers is not in the document]
Comment (Sagsgit) [reply microsoft365/microsoft365.py]: fixed

File: microsoft365/microsoft365.py [hunk 2]
```diff
@@ -2210,4 +2210,430 @@ async def execute(self, inputs: Dict[str, Any], context: ExecutionContext) -> Ac
...
+
+            for schedule in response.get("value", []):
+                email = schedule.get("scheduleId", "")
+                schedule_items = schedule.get("scheduleItems", [])
+
+                # Check for conflicts (any non-free items)
+                conflicts = []
+                for item in schedule_items:
+                    status = item.get("status", "")
+                    if status in ("busy", "tentative", "oof", "workingElsewhere"):
```
[the remark this answers is not in the document]
Comment (Sagsgit) [reply microsoft365/microsoft365.py]: fixed

**Comment by Sagsgit** [comment]
@codex?

**Comment by vish4004** [comment]
@codex can you review these points and give your opinion

Summary of Required Changes

  1. Remove meeting_duration_preference from README (or implement it) — doc/code mismatch
  2. Replace datetime.utcnow() with datetime.now(timezone.utc) — deprecated in Python 3.12+
  3. Move from datetime import ... to top of file
  4. Remove redundant if time_constraint: check
  5. Add trailing newline to microsoft365.py

  Suggested Improvements (non-blocking)

  6. Verify whether Schedule.Read.All scope is actually needed given Calendars.ReadWrite is present
  7. Add pagination support to list_rooms
  8. Add test for default date handling in find_meeting_times
  9. Separate formatting-only config.json changes into their own commit

**Review by vish4004 — COMMENTED** [review]
# PR #168 Review: Microsoft 365 Meeting Scheduling & Room Management

## Overall Assessment: **Approve with requested changes**

The PR is well-structured and follows existing patterns closely. The four new actions (`find_meeting_times`, `get_schedule`, `list_rooms`, `check_room_availability`) are properly implemented against the MS Graph API. However, there are several issues ranging from a documentation bug to a deprecated API usage that should be addressed.

---

## CRITICAL Issues

### 1. `meeting_duration_preference` parameter documented but not implemented
The README documents a `meeting_duration_preference` parameter with values `morning, afternoon, evening` for `find_meeting_times`, but:
- It's **missing from `config.json` input_schema** (not in the properties list)
- It's **not implemented in the Python code**

This maps to the `activityDomain` property in the MS Graph `timeConstraint` object (values: `work`, `personal`, `unrestricted`). However, "morning/afternoon/evening" doesn't directly map to any Graph API parameter — it would need custom time constraint logic.

**Fix:** Either implement it or remove it from the README. Removing is simpler since `start_datetime`/`end_datetime` already allow callers to constrain the time range.

### 2. `datetime.utcnow()` is deprecated (Python 3.12+)
In `find_meeting_times` (`microsoft365.py`):
```python
start_parsed = datetime.utcnow()
```
Since the SDK requires Python 3.13+, this should be:
```python
from datetime import datetime, timedelta, timezone
start_parsed = datetime.now(timezone.utc)
```

---

## MEDIUM Issues

### 3. Scope additions may be unnecessarily broad
The PR adds `Schedule.Read.All` and `Place.Read.All` to the OAuth scopes.

- **`Place.Read.All`** — Correct and required for the Places API (`list_rooms`). ✅
- **`Schedule.Read.All`** — Per MS Graph docs, `getSchedule` only requires `Calendars.ReadBasic` (delegated) or `Calendars.Read` (app), and `findMeetingTimes` requires `Calendars.Read.Shared`. The integration already has `Calendars.ReadWrite`, which is a superset of these.

`Schedule.Read.All` may be redundant. Adding unnecessary scopes widens the permission surface area, which some IT admins will flag during consent review.

**Recommendation:** Verify whether `Calendars.ReadWrite` (already present) covers the getSchedule and findMeetingTimes use cases in practice. If so, remove `Schedule.Read.All`.

### 4. No pagination for `list_rooms`
The MS Graph Places API supports pagination via `@odata.nextLink`. The existing actions in this integration (e.g., `list_emails`, `list_files`) handle pagination by recursively following nextLink. The `list_rooms` action does not — it fetches only a single page. Organizations with many rooms would get silently truncated results.

**Recommendation:** Add nextLink pagination handling consistent with existing actions, or at minimum document the limitation.

### 5. `from datetime import ...` inside try block
In `find_meeting_times`, the import is inside the function body:
```python
try:
    ...
    from datetime import datetime, timedelta
```
This should be at the top of the file with other imports. While Python caches imports so it's not a performance issue, it's inconsistent with the rest of the codebase.

### 6. Redundant `if time_constraint:` check
```python
if time_constraint:
    body["timeConstraint"] = time_constraint
```
`time_constraint` is always set (either from user-provided dates or defaults), so this condition is always True. Remove the conditional and just assign it directly.

---

## MINOR Issues

### 7. No newline at end of `microsoft365.py`
The diff shows `\ No newline at end of file` for the implementation file. Add a trailing newline.

### 8. Massive formatting changes mixed with feature changes
A large portion of the config.json diff (~500+ lines) is just reformatting existing JSON from compact to expanded multi-line format. While the expanded format is arguably more readable, mixing formatting changes with feature changes makes the PR harder to review. Ideally these would be separate commits.

### 9. Hardcoded `availabilityViewInterval: 15` in `check_room_availability`
The interval is hardcoded to 15 minutes. This is reasonable for room checking but could be documented (why 15 specifically) or made configurable.

---

## SDK Compliance Check

| Pattern | Status | Notes |
|---|---|---|
| Extends `ActionHandler` | ✅ | All 4 new handlers properly extend `ActionHandler` |
| Returns `ActionResult` | ✅ | Consistent with existing actions |
| `@microsoft365.action()` decorator | ✅ | All properly decorated |
| Error handling pattern | ✅ | try/except returning `result: False` with error string |
| Uses `context.fetch()` | ✅ | Proper use of SDK HTTP client |
| `cost_usd=0.0` | ✅ | Consistent with existing actions |
| `config.json` schema | ✅ | input_schema and output_schema properly defined |
| `display_name` on actions | ✅ | All 4 actions have display_name |

---

## MS Graph API Compliance Check

| Endpoint | Status | Notes |
|---|---|---|
| `POST /me/findMeetingTimes` | ✅ | Correct URL, correct request body structure |
| `meetingDuration` format (`PT{n}M`) | ✅ | ISO 8601 duration correctly used |
| `timeConstraint.timeslots` structure | ✅ | dateTime + timeZone objects correct |
| `locationConstraint` structure | ✅ | `resolveAvailability`, `locationEmailAddress` correct |
| `POST /me/calendar/getSchedule` | ✅ | Correct URL, correct request body |
| `startTime`/`endTime` structure | ✅ | dateTime + timeZone objects correct |
| `GET /places/microsoft.graph.room` | ✅ | Correct URL for all rooms |
| `GET /places/microsoft.graph.roomList` | ✅ | Correct URL for room lists |
| `GET /places/{email}/microsoft.graph.roomList/rooms` | ✅ | Correct URL for rooms in list (uses email, not ID) |
| Response field mappings | ✅ | `scheduleId`, `availabilityView`, `scheduleItems`, `workingHours` all correctly mapped |
| `isWheelChairAccessible` casing | ✅ | Matches Graph API property name |

---

## Test Coverage Assessment

| Test | Covered |
|---|---|
| find_meeting_times - success | ✅ |
| find_meeting_times - no suggestions | ✅ |
| find_meeting_times - location constraint | ✅ |
| find_meeting_times - error | ✅ |
| find_meeting_times - default dates (no start/end) | ❌ Missing |
| find_meeting_times - 7-digit fractional seconds | ❌ Missing |
| get_schedule - success with working hours | ✅ |
| get_schedule - error | ✅ |
| get_schedule - per-schedule errors | ❌ Missing |
| list_rooms - all rooms | ✅ |
| list_rooms - room lists | ✅ |
| list_rooms - rooms in list | ✅ |
| list_rooms - missing email error | ✅ |
| list_rooms - API error | ✅ |
| check_room_availability - all available | ✅ |
| check_room_availability - mixed conflicts | ✅ |
| check_room_availability - error | ✅ |

**Coverage is good** (14 tests covering success, error, and edge cases), but there are a few gaps around default date handling and fractional-second timestamp parsing (which the code explicitly handles but doesn't test).

---

## Summary of Required Changes

1. **Remove `meeting_duration_preference` from README** (or implement it) — doc/code mismatch
2. **Replace `datetime.utcnow()` with `datetime.now(timezone.utc)`** — deprecated in Python 3.12+
3. **Move `from datetime import ...` to top of file**
4. **Remove redundant `if time_constraint:` check**
5. **Add trailing newline to `microsoft365.py`**

## Suggested Improvements (non-blocking)

6. Verify whether `Schedule.Read.All` scope is actually needed given `Calendars.ReadWrite` is present
7. Add pagination support to `list_rooms`
8. Add test for default date handling in `find_meeting_times`
9. Separate formatting-only config.json changes into their own commit

File: microsoft365/microsoft365.py [hunk 3]
```diff
@@ -2210,4 +2210,446 @@ async def execute(self, inputs: Dict[str, Any], context: ExecutionContext) -> Ac
...
+    async def execute(self, inputs: Dict[str, Any], context: ExecutionContext) -> ActionResult:
+        try:
+            list_type = inputs.get("list_type", "rooms")
+            limit = inputs.get("limit", 100)
+
+            if list_type == "room_lists":
+                # GET /places/microsoft.graph.roomList
+                url = f"{GRAPH_API_BASE}/places/microsoft.graph.roomList"
+                params = {"$top": limit}
+                response = await context.fetch(url, params=params)
```
[the remark this answers is not in the document]
Comment (Sagsgit) [reply microsoft365/microsoft365.py]: fixed this in this https://github.com/Autohive-AI/autohive-integrations/pull/168/commits/fb2fac2379d4339dd59d83da3481e916e300b777

**Comment by Sagsgit** [comment]
@codex

**Comment by phillip-haydon** [comment]
# 🤖 Automated Code Review - Mrs. Reviewer

## Executive Summary

This PR adds intelligent meeting scheduling and room management capabilities to the Microsoft 365 integration. The code quality is **generally good** with proper async patterns, bounded pagination, and appropriate error handling. However, there are **code duplication issues** in the `list_rooms` action that should be addressed for better maintainability.

**Overall Verdict:** ✅ **Patch is correct** - No blocking issues found. The code will not break existing functionality and is free of critical bugs.

---

## 📋 Review Findings

### [P2] Code Duplication in list_rooms Pagination Logic
**File:** `microsoft365/microsoft365.py`  
**Lines:** 2424-2541  
**Confidence:** 0.95

The `ListRoomsAction` contains three nearly identical pagination blocks for `room_lists`, `rooms_in_list`, and default `rooms`. While this doesn't cause runtime issues, it increases maintenance burden and risk of inconsistent behavior if one branch is updated but others aren't.

**Current pattern (repeated 3 times):**
```python
url = f"{GRAPH_API_BASE}/places/microsoft.graph.roomList"
params = {"$top": limit}
all_items = []
next_url = url
is_first = True
while next_url and len(all_items) < limit:
    response = await context.fetch(next_url, params=params if is_first else None)
    is_first = False
    all_items.extend(response.get("value", []))
    next_url = response.get("@odata.nextLink")
all_items = all_items[:limit]
```

**Recommended refactor:**
```python
async def _fetch_paginated(context, url: str, limit: int) -> List[Dict]:
    """Helper to fetch paginated results from Graph API."""
    params = {"$top": limit}
    all_items = []
    next_url = url
    is_first = True
    while next_url and len(all_items) < limit:
        response = await context.fetch(next_url, params=params if is_first else None)
        is_first = False
        all_items.extend(response.get("value", []))
        next_url = response.get("@odata.nextLink")
    return all_items[:limit]
```

---

### [P2] Repeated Room Transformation Logic
**File:** `microsoft365/microsoft365.py`  
**Lines:** 2448-2461, 2477-2498, 2515-2536  
**Confidence:** 0.90

The room transformation logic (extracting `id`, `display_name`, `email_address`, etc.) is duplicated between `rooms_in_list` and default `rooms` branches with identical field mappings.

**Recommendation:** Extract to a `_transform_room(room: Dict) -> Dict` helper function to follow DRY principles.

---

## ✅ Positive Patterns Observed

The following patterns are well-implemented:

1. **✅ Async/await used correctly** - All API calls use `await context.fetch()` properly, avoiding blocking operations.

2. **✅ Pagination with early termination** - Pagination loops correctly check `len(all_items) < limit` to avoid fetching more data than needed.

3. **✅ Batch API operations** - `get_schedule` and `check_room_availability` batch multiple email addresses into a single API call rather than making N+1 requests.

4. **✅ Reasonable defaults** - `max_candidates` is capped at 20 to prevent unbounded API responses.

5. **✅ Datetime parsing optimization** - The fractional seconds truncation (`frac[:6]`) is a reasonable approach to handle Microsoft Graph's 7-digit precision.

6. **✅ Bounded memory usage** - All pagination is properly bounded by user-specified limits (default 100), preventing unbounded memory growth.

7. **✅ Python 3.12+ compatibility** - Correctly updated `datetime.utcnow()` to `datetime.now(timezone.utc)`.

---

## 🔍 No Issues Found In

- **Security:** Input validation, OAuth2 scopes, error handling
- **Memory Management:** No leaks, proper bounded pagination
- **Performance:** Efficient async patterns, appropriate batching
- **Error Handling:** Comprehensive try-catch blocks with proper error responses

---

## 📊 Summary Table

| Category | Status | Notes |
|----------|--------|-------|
| **Correctness** | ✅ Pass | No blocking bugs |
| **Security** | ✅ Pass | Proper OAuth2 scopes, input validation |
| **Performance** | ✅ Pass | Efficient async patterns |
| **Memory** | ✅ Pass | Bounded pagination, no leaks |
| **Maintainability** | ⚠️ Minor Issues | Code duplication in list_rooms |

---

## 🎯 Recommendations

**Priority 2 (Normal - Fix Eventually):**
1. Refactor `list_rooms` pagination logic into a shared helper function
2. Extract room transformation logic into a reusable function

These are **maintainability improvements** rather than bugs - they don't affect runtime behavior but will make future maintenance easier.

---

## 🏆 Overall Assessment

**The patch is correct.** ✅

The implementation is solid with good async patterns, proper error handling, and bounded resource usage. The code duplication issues are minor maintainability concerns that don't impact functionality. The PR successfully adds the meeting scheduling features without introducing bugs or breaking changes.

---

*This is an automated review from **Mrs. Reviewer** in [Autohive](https://www.autohive.com). Powered by specialized AI review agents for General, Security, Performance, and Memory analysis.*

**Review by vish4004 — DISMISSED** [review]
Looks good, thanks

File: microsoft365/config.json [hunk 4]
```diff
@@ -1,7 +1,7 @@
 {
     "name": "Microsoft Copilot 365",
     "version": "1.0.0",
```
Comment (TheRealAgentK) [inline microsoft365/config.json]: Version should probably change with a new release.
